snippets/csharp/System.Xml/XmlDocument/Overview/xmlhelpermethods.cs

```diff
@@ -98,35 +98,24 @@ public XmlDocument LoadDocumentWithSchemaValidation(bool generateXML, bool gener
             // Helper method to retrieve schema.
             XmlSchema schema = getSchema(generateSchema);
 
-            if (schema == null)
-            {
-                return null;
-            }
-
             settings.Schemas.Add(schema);
 
             settings.ValidationEventHandler += ValidationCallback;
             settings.ValidationFlags =
                 settings.ValidationFlags | XmlSchemaValidationFlags.ReportValidationWarnings;
             settings.ValidationType = ValidationType.Schema;
 
-            try
+            if (generateXML)
             {
-                reader = XmlReader.Create("booksData.xml", settings);
+                string xml = generateXMLString();
+                StringReader stringReader = new StringReader(xml);
+
+                reader = XmlReader.Create(stringReader, settings);
             }
-            catch (System.IO.FileNotFoundException)
```

Comment:
Is it okay to remove handling of FileNotFoundException? XmlReader.Create could throw it if file does not exist. The guidelines  say that samples should catch common exceptions that could be thrown by the method: https://docs.microsoft.com/en-us/contribute/dotnet/dotnet-contribute#write-a-new-sample

>Your sample should include appropriate exception handling. It should handle all exceptions that are likely to be thrown in the context of the sample.

Reply:
I am open to criticism but I'm not seeing a clean exception handling example coming out of this code.

The original code did not handle the exception. The original code returned null on that one error (FileNotFoundException) and all other errors such as malformed XML threw an exception. On a file access issue threw an exception. It was inconsistent error handling. Returning value errors is 1985 C programming style. The original is an example of incorrect error handling.

generateSchema and generateXML are required parameters but are only used when a FileNotFoundException is thrown. There names appear to me "generate" but in reality they should be generateIfTheHardCodeFileIsNotFound. Not a clear method signature.

A developer who wanted to handle file not found would use File.Exists so as not to generate an FileNotFoundException.

I don't see a clear set of common exceptions. Four of the five documented Exceptions for XmlReader.Create are possible. There are a half dozen other methods that could also generate errors.

I could put a try/catch and a Console.Error.Writeline but I don't see it as a realistic example of error.

Once this change is approved I can do another one and get a "using" to take advantage of XmlReader's IDisposable. That is a lessen that should be taught.

Reply:
>I am open to criticism but I'm not seeing a clean exception handling example coming out of this code.

The code currently tries to open xml file for reading, and if it throws exception, instead fallbacks to some hardcoded xml. I'm not sure is it clean or not, but you PR seems to change the meaning of the code, to make it throw where it previously wasn't throwing. It's not just a cleanup. So i wonder whether it's OK to do it in this case.

>On a file access issue threw an exception. It was inconsistent error handling.

You are right. In fact, the guidelines i cited aren't followed consistently, because it's not always easy to determine "likely" exceptions. Maybe this sample needs some improvement, like handle IOException instead to account for more error conditions. I don't see it as an argument to completely remove the exception handling, though.

>Returning value errors is 1985 C programming style. The original is an example of incorrect error handling.

I have no idea what do you mean by this. Such pattern is used all the time in .NET. The production code might need to log the exception additionally when catching it, but for an example, this error handling seems fine.

Reply:
@MSDN-WhiteKnight how about this?

if generateXML==true and the XML file is missing, catch the FileNotFoundException and fall back on the hard-coded XML.

if generateSchema ==true and the schema file is missing, catch the FileNotFoundException and fall back on the hard-coded schema.

I think it has a better workflow than the original.

Reply:
I can add the using to the XmlReader. I am not sure how Microsoft wants their code written. Do we go old school?

using (Xmlreader reader = XmlReader.Create(blah blah))
{
}

Or new school?
using Xmlreader reader = XmlReader.Create(blah blah);

Reply:
Fine to use new language features. I'm not Microsoft, but using declaration syntax is already used, for example, here: https://docs.microsoft.com/en-us/dotnet/api/system.reflection.metadata.metadatareader?view=net-6.0#examples